### PR TITLE
simplify some async fixtures

### PR DESCRIPTION
**Pull request by minrk**
use async fixtures for simpler event-loop integration

several of these fixtures were written before fixtures themselves could be async, but now they can, which means we can use async/await instead of run_sync.

related to https://github.com/jupyterhub/pytest-jupyterhub/issues/30

### CODE REVIEW

**Comment by GeorgianaElena** [comment]
@Sheila-nk, do you want to help out with reviewing Min's PR?

I believe it could help with putting the info in https://github.com/jupyterhub/pytest-jupyterhub/issues/30 into context, but also identify what are the confusion points, so that we can document them. Feel free to ping and ask any additional clarification. Asking questions is an important piece in the review process.

What do you think?

**Comment by minrk** [comment]
I tried to remove the io_loop fixture, which I don't believe we _really_ need anymore, but it did cause some failures. The only thing the io_loop fixture really does is call [`io_loop.close(all_fds=True)`](https://github.com/jupyterhub/jupyterhub/blob/e85b91cd9bc24db6e09d0c6d17f1e754dc08c72c/jupyterhub/tests/conftest.py#L140) to ensure everything's cleaned up and FDs don't leak from one test module to the next. This isn't necessary if we actually clean up everything, which we _mostly_ do. But I did see failures such that at least _sometimes_ we rely on this to close the http socket for the server, so I decided to leave it alone for now. I think if the `app` fixture more reliably cleaned up everything it started, this would not be necessary.

**Comment by Sheila-nk** [comment]
@minrk @GeorgianaElena 👋 I have some questions first.

> In JupyterHub, we use tornado, which is a wrapper around asyncio's event loop. It's unlikely that the tornado io_loop will be needed directly in tests. It's very important that the tornado io loop is 'hooked up to' the main asyncio event loop, otherwise you'll get mysterious hangs, as events you are waiting for just never happen because they are hooked up to an event loop that's not running.

I was trying to understand why the `io_loop` is no longer needed considering JupyterHub uses tornado as you had mentioned. When you say that tornado is a wrapper around `asyncio` does that mean that pytest-asyncio can be used to write tests for Tornado applications? And is 'hooking' to the main event loop necessary anymore?

Second,  why is this `io_loop` still calling the `event_loop` fixture even with the `async` keyword provided? Is it necessary because of the override?

Lastly, from studying the changes to the `app` fixture, my understanding is that the `make_app` function had to go since the `initialize` and `start` functions are already registered coroutines, so they can easily be called using the `await` keyword within the `app` coroutine. Am I correct in saying this?

**Comment by minrk** [comment]
> I was trying to understand why the io_loop is no longer needed considering JupyterHub uses tornado as you had mentioned

To be clear, the tornado IOLoop _object_ is still used. What may not be needed is the _fixture_ to do special setup and teardown of the loop. Instantiating JupyterHub calls `IOLoop.current()` which gets the IOLoop associated with the current asyncio event loop, and creates it if necessary. _If_ the MockHub cleanup fully closes and stops everything, the extra io_loop teardown doesn't need to happen. That doesn't appear to be _quite_ always the case.

> When you say that tornado is a wrapper around asyncio does that mean that pytest-asyncio can be used to write tests for Tornado applications?

Yes, there's not generally any need for special handling of tornado _unless_ you also want/need to _close_ the IOLoop's resources at the end. Tornado calls will generally work fine if run from inside asyncio without doing any special tornado setup beforehand. This has changed since we started writing the jupyterhub test suite, and there are still some remnants of the old ways. It's much simpler now.

>  And is 'hooking' to the main event loop necessary anymore?

The 'hooking' is done in tornado itself, we mainly assert that it happens, just in case something goes wrong in tornado setup which would cause the two-loops-problem. Calling `AsyncIOMainLoop()` from inside a coroutine (why we `async def io_loop()`) will always create an IOLoop attached to the _current_ event_loop.

> Second, why is this io_loop still calling the event_loop fixture even with the async keyword provided? Is it necessary because of the override?

Only for the assertion, really. We could remove it an assert against `asyncio.get_running_loop()`. But putting the reference there also ensures that setup/teardown happens in the right order.

> from studying the changes to the app fixture, my understanding is that the make_app function had to go since the initialize and start functions are already registered coroutines, so they can easily be called using the await keyword within the app coroutine. Am I correct in saying this?

Pretty much. I wouldn't say it _had to go_, but it _could_ go because we can now use top-level awaits. We only had the function because we need functions to pass to `run_sync`. Since moving to `async def` means we can have top-level awaits, we no longer need that extra layer for wrapping `await`s, and can do the steps we wanted in order in a simpler way without a wrapper.

File: jupyterhub/tests/conftest.py [hunk 1]
```diff
@@ -322,26 +325,26 @@ def cleanup():
...
+            await service.server.wait_up(http=True)
     return service
 
 
 @fixture
-def mockservice(request, app):
+async def mockservice(request, app):
     """Mock a service with no external service url"""
-    yield _mockservice(request, app, url=False)
+    yield await _mockservice(request, app, url=False)
```
Comment (Sheila-nk) [inline jupyterhub/tests/conftest.py:335]: I have not encountered a scenario where `yield` and `await` are used in this way. Can you please help me understand the usage?

Comment (Sheila-nk) [inline jupyterhub/tests/conftest.py:335]: I have not encountered a scenario where `yield` and `await` are used in this way. Can you please help me understand the usage?

Comment (minrk) [reply jupyterhub/tests/conftest.py:335]: Absolutely! Fixtures are used to do one or both of:

1. set something up before the test
2. clean up after the test

If you only need to do setup, the fixture function can `return` what should be the value of the fixture (if any) in tests that use it.

If you need to _cleanup_, then you can `yield` what should be passed to the functions (can be nothing, if it's purely setup/teardown and has no object to be passed to tests). You can read more in the [pytest docs on fixture teardown/cleanup](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended).

So a fixture with setup _and_ teardown generally looks like:

```python
@pytest.fixture
def my_fixture():
    setup_before()
    yield value_for_tests_if_any
    cleanup_after()
```

The `await` doesn't change anything here. When you `yield <expression>`, what's yielded is the _result_ of the expression, which means the `awaited` result if there's an `await` in there. That means

```python
yield await async_thing
```

is the same as

```python
thing = await async_thing
yield thing
```

if that makes things clearer.


If you run [this sample test](https://gist.github.com/d6fa0aaaa35ec8bad55db5136499c728) with:

```
pytest -vs example.py
```

you'll get:

```
example.py::test_sample in setup
in async function
myfixture='sync result'
PASSEDin teardown
```

Comment (Sheila-nk) [reply jupyterhub/tests/conftest.py:335]: I definitely understand this better. Thank you for the explanation.
When I run the code I get this though:
```python
example.py::test_sample <async_generator object myfixture at 0x7f22b8408af0>
PASSED
```

Comment (minrk) [reply jupyterhub/tests/conftest.py:335]: Forgot about my global config. It should be:

```
pytest -vs --asyncio-mode auto example.py

Comment (Sheila-nk) [reply jupyterhub/tests/conftest.py:335]: Yeah! This works now. Thank you!

**Comment by Sheila-nk** [comment]
@minrk Thank you for taking the time to answer my questions. Getting to review this PR was insightful.
 I think this PR is ready for merging 🚀 
@GeorgianaElena I will leave that to you

**Comment by GeorgianaElena** [comment]
Thanks @Sheila-nk and @minrk

**Pull request merged**